### ISSUE TITLE
journal *all* the things? (v10)

### ISSUE THREAD

**Issue by rsek** [ben/foundry-ironsworn#386]
so, there's a lot of new journal features in v10. i have this hypothesis that most or all things that are currently "actors" (aside from PCs) could be journal entries, too, so i'd like to do a brain dump to see if it still seems clever once i read it back ;)

## Possible benefits
* leverage FVTT's improved infrastructure for creating, organizing, searching, and storing journal entries (better features, extensible with other modules, easier adoption by FVTT-savvy players)
* fewer dependencies + less time spent troubleshooting the WYSIWIG editor package
* make sheet *display* more modular: currently you need the *whole dang sheet* in order to get any sheet information. a PC sheet with progress tracks offloaded to 
* reduce amount of confusing, overlapping functionality
* semantically logical within the "gaming table" metaphor. put another most actors represent something you're *making notes about* in ironsworn. where do i make notes? the journal! going to "actors" to create a location is less than 

## vs. Current Actor Functionality
to figure out what these new journal entry types should be able to do, it's worth thinking what the current actor-based approach does first. what functionality could a journal-based approach provide

### Guiding the player in entity creation
i think there's a lot of value in having a scaffolding/template/form so that people don't have to look up "what's the recommended way to generate this again?"

**Possible journal-based approach:** could be accomplished in much the same way - multiple journal entry types or templates. user picks on journal entry creation.

### Curate a manageable subset of oracle/table shortcuts
the current location builders do this, and it's really useful, but people can only use what's hard-coded in.

i'd file Delve matrices and Site Themes/Domains under this heading. they're structured a bit differently, but their function is similar.

**Possible journal-based approach:** include shortcuts to roll oracles. this might be via a custom UI element containing a bunch of buttons. ideally, users would be able to customize this themselves. it might also be used used as the basis for oracle arrays?

### Curate a manageable subset of moves

i haven't handled the delve module much - does it do this with the Delve moves? another place it might be useful is combat.

### Instantly roll a bunch of oracles
e.g. SF locations.

**Possible journal-based approach:** journal entry can include a "preset" of a group of oracles to roll with a single click?

### Associate tracks/clocks with an entity
see #382 for some thoughts on that. making a progress track and clock independently-toggleable items on journal entries could account for this.

### Create a representation of the game entity on the virtual tabletop
currently, this is accomplished by actor tokens. IMO actors are *way* more complicated than any non-PC entity needs to be in ironsworn. 

**Possible journal-based approach:** map pins are getting a refresh in v10. they have an icon, a tooltip, and can be linked to specific journal entries - more than sufficient unless someone needs a battlemap (which, IMO, is out of scope for a game whose default setting is "theatre of the mind", and a pretty substantial divergence from presumed gameplay).

### Make the associated content easily findable by other PCs
pretty straightforward translation here. they'll just be in the journal tab (and searchable!) rather than the actor tab.

## Putting it all together

at this point, i'm imagining something like this:

* each of the individual functions above extends the base JournalEntry like discrete "plugins"
* each "plugin" is modular - as self-contained as possible within a single child object of JournalEntryData
* `foundry-ironsworn` provides preset journal entry "templates" corresponding to previous actor types
* users can toggle any of the JournalEntry "plugin" on or off to tweak their template as desired
* bonus points: users can save their custom arrangement as a template. def not an MVP! but something to think about when designing the initial offering
* extra bonus points: some means for users to share and import these journal templates?

**Comment by rsek** [comment]
some of the issues relevant to this:
* https://github.com/foundryvtt/foundryvtt/issues/4941
* https://github.com/foundryvtt/foundryvtt/issues/7236
* https://github.com/foundryvtt/foundryvtt/issues/7201
* https://github.com/foundryvtt/foundryvtt/issues/7236
* https://github.com/foundryvtt/foundryvtt/issues/7050
* https://github.com/foundryvtt/foundryvtt/issues/6945

**Comment by rsek** [comment]
started digging around in v10's source, so here's some more notes.

`JournalEntry` is basically a container with a header a bunch of child `JournalEntryPages`. `JournalEntryPages` can be viewed one at a time, or collated into a single long text with subheaders. they're organized into an outline

<img width="613" alt="Screen Shot 2022-06-11 at 3 59 28 AM" src="https://user-images.githubusercontent.com/5354757/173185096-9e9bcead-14a4-43a6-b145-a1d1751f089a.png">

the "canonical" types of journal entry pages.
<img width="374" alt="Screen Shot 2022-06-11 at 4 02 14 AM" src="https://user-images.githubusercontent.com/5354757/173185183-add4f152-0085-4394-8f94-0df2368574cc.png">

it seems those types can be expanded by the system.
<img width="575" alt="Screen Shot 2022-06-11 at 3 54 52 AM" src="https://user-images.githubusercontent.com/5354757/173184915-51274217-1388-4b47-afa7-0d601dc0420f.png">

markdown and html are both options for page text. nice.
<img width="386" alt="Screen Shot 2022-06-11 at 4 08 53 AM" src="https://user-images.githubusercontent.com/5354757/173185368-222e33e4-6a6e-4a32-9822-cfffe0717bfe.png">

dropping a journal page on to the map creates a map pin, as one might expect. the icon can now be set to anything.
<img width="703" alt="Screen Shot 2022-06-11 at 4 26 15 AM" src="https://user-images.githubusercontent.com/5354757/173185892-61b398c2-17a5-4d3e-9f51-2dbe7289690c.png">

currently it doesn't provide any way of configuring/hiding the square ground, though. 🤔  i've written up an issue for it: https://github.com/foundryvtt/foundryvtt/issues/7255

**Comment by ben** [comment]
Some thoughts, in no particular order.

**I like the idea of having the JE be the source of truth for progress "items".** It removes the concept of actor ownership, so now two actors could have shortcuts to a single Vow on their sheets, and not get out of sync.

**I still think there's value in having these tabs in the character sheet.** I still want it to be possible to play using a single sheet, and also to have a "dashboard" of your current progress at a glance. I'm fine with popping out new apps for notes and changing parameters, but overview information and simple "mark progress" controls should be on the main sheet.

<img src="https://user-images.githubusercontent.com/39902/173193291-636374b4-c4d1-44a5-93c2-e629d170cde1.jpg" width="400" />

**Putting JE's on the map is still clunky.** This might change after v10dev1, but out of the box the experience is pretty bad for this:
- Dragging a JE to the map activates the "bookmark" icon, which means you're now in a different mode and can't interact with the actors on the map
- Switching back to "actor" mode hides the JE pins
- Once you realize you have to activate the "pin" icon in bookmark mode, you still can't drag JE's in actor mode, and you can't interact with actors at all in bookmark mode
- …which means you're switching back and forth between map nodes when you want to move these things around, and you have to remember which ones are which type. Do not want.

I think this breaks a lot of flows, so it still seems to me that actors are the way to go. NPC actors will still be just a thin shell around a "progress item" (i.e. a JE), but actors they should be.

**Augmenting the JE UI could replace the current progress sheet.** Even without the v10 JE fancy-pants stuff, we could still just pop a button into the title bar that opens a sidebar. A quick mockup:
![CleanShot 2022-06-11 at 08 13 42](https://user-images.githubusercontent.com/39902/173193678-e296e579-f802-41c4-ac1c-b0a762c0fd11.jpg)

Again, the progress/rank/clock/type data is stored on the JE itself, and references to that JE are spread around to the actors that care about it.

**Comment by DiceT** [comment]
> - Dragging a JE to the map activates the "bookmark" icon, which means you're now in a different mode and can't interact with the actors on the map
> - Switching back to "actor" mode hides the JE pins
> - Once you realize you have to activate the "pin" icon in bookmark mode, you still can't drag JE's in actor mode, and you can't interact with actors at all in bookmark mode
> - …which means you're switching back and forth between map nodes when you want to move these things around, and you have to remember which ones are which type. Do not want.

There are ways to enable the Journal layer to be available when viewing other layers. Someone has done this, though I do not recall which system/module.

**Comment by rsek** [comment]
> I think this breaks a lot of flows, so it still seems to me that actors are the way to go. NPC actors will still be just a thin shell around a "progress item" (i.e. a JE), but actors they should be.

hmm, reckon you could elaborate a bit on this? like... what functionality does keeping them as actors provide that's useful *for ironsworn*?

moving actors around on a battlemap is one thing, but battlemaps aren't encouraged or even suggested by IS/SF. in essence, it'd mean complicating the system by making one non-player game entity an actor and *everything else* a journal entry to enable niche variations of gameplay. people going "off-road" in that regard are probably better served by things outside the scope of this game system, like other modules.

alternatively: offer an *option* to generate a linked actor token from a journal entry with a progress track, but IMO it shouldn't be the default, because it's not the default position of the game itself, and adding optional complexity is the realm of "advanced options". but, frankly? i wouldn't even bother unless someone specifically requested it. like, in a journal-entries-as-the-primary-non-player-entity

and as much as the semantics of "actor = npc" are slightly more obvious - functionally speaking, PCs are the only "actors" in ironsworn anyways. nobody else makes moves.

as far as the convenience angle: i mean, fair, but not insurmountable. i think @DiceT has something there, and i reckon there's worse ideas than adding a dependency to make the map/journal experience smoother. IMO "having to go look for modules to improve the experience when playing the game roughly as-written" is itself a UX flaw. if it's possible to remedy that with a dependency, then that's a fairly easy problem to solve (though, ideally - make it something folks can toggle so they can use alternative approaches if desired). by leaving the stock map handling in there, the system is - functionally speaking - shipping with a mapping UX that isn't "fit for purpose".

**Comment by rsek** [comment]
oh, btw. we wanted a bespoke map-note/placeable solution instead of relying on someone else's code, the API is getting some stuff that could make that easier:

* https://github.com/foundryvtt/foundryvtt/issues/6475
* https://github.com/foundryvtt/foundryvtt/issues/7007

building a whole custom layer might be out of scope though ;) IMO an add'l module depedency is good enough to get a working prototype online.

also - maybe it'd be worth opening a feature request with FVTT to improve the awkward map note workflow out of the box? (i'd do it myself, but i think @ben has a better understanding of its limitations than i do, lol). i reckon the crux of the issue is a mismatch between the expectations of tactical gameplay and setting-level maps; improving it could make the VTT more versatile for non-tactical-battlemap mapping styles. and if there's a time to get it on their radar, it's the update where they're overhauling this stuff!

**Comment by rsek** [comment]
re: this system including a more "opinionated" default map configuration... on one hand, embedding a highly opinionated map experienc in the module might be some overreach. but on the other hand: foundry already *has* a highly opinionated map experience out of the box - it's just one with an opinion that doesn't match up with ironsworn's presumed gameplay. so we might need to be opinionated just to cancel that out, ha.

possibilities for prototyping a v10-like map experience on v9, or offering backwards compatibility:
* https://www.foundryvtt-hub.com/package/pin-cushion/ - set any icon
* remove pin background: https://www.foundryvtt-hub.com/package/backgroundless-pins/ (v10 will need this if [this](https://github.com/foundryvtt/foundryvtt/issues/7255) doesn't make it)

QoL improvements:
* pincushion, above, provides some useful functions beyond setting custom icons:
  * Adds the ability to double-click the canvas while on the Notes Layer and create a map pin (and corresponding Journal Entry)
  * Adds a preview of the associated Journal Entry when you hover over a map pin
* drop a whole folder on to the map. might be leveraged for building a sector?: https://www.foundryvtt-hub.com/package/dfreds-droppables/
* quick insert appears to be invocable from any layer, and can return journal entries (presumably for use as map notes)? https://www.foundryvtt-hub.com/package/quick-insert/

**Comment by ben** [comment]
> offer an _option_ to generate a linked actor token from a journal entry with a progress track, but IMO it shouldn't be the default, because it's not the default position of the game itself

Yes, exactly. The _default_ flow would be "open your PC sheet, click the `+` on the progress/connections tab, and that generates a JE and a linkage to this PC. But there can be a button on the JE sidebar that does a create-or-open on an actor (which also references that same JE), and makes it placeable on the map. This seems to be hard to explain in words, but I have a fairly clear picture in my mind.

> hmm, reckon you could elaborate a bit on this? like... what functionality does keeping them as actors provide that's useful _for ironsworn_?

See #121 and  #238 for the original impetus here, but there have been quite a few requests over the lifetime of this project. Especially when these items can represent named NPCs, it sometimes makes sense to plop them onto the map to remind you who your contact is on this planet, or where Admiral Carcosa's fleet is now. The use case isn't "gridded map for tactical combat," it's more "desktop that holds the minimized items that you're working with."

> PCs are the only "actors" in ironsworn anyways. nobody else makes moves.

I don't think anyone is going to interpret the Foundry "actors" tab in that literal a sense. I learned on my first day with Foundry that an `actor` is "a thing you can put on the map."

Regarding Foundry's opinionated division of JE's and actors on the map: the thing they're trying to enable is the "keyed map with location notes" experience. In D&D, when you're dragging PC tokens around, you _absolutely_ don't want to drag the note about the ballroom into the adjoining torture chamber. The interaction model is exactly what they want it to be, and none of the modules you've linked alter this behavior, so it seems like they're fit for their purpose. My overall opinion is that trying to make map pins behave like actor tokens is a lot of swimming upstream, changing a core Foundry feature in ways that will upset those player who _want_ the default experience, and will involve a bunch of effort and bugs for not much advantage over "generate an actor and link it to the JE."

Regarding all the modules you've found: I think most of these fall into the "this is a niche that I'm glad is supported by a module, but isn't core to the experience" category, and I don't think we need to depend on or reimplement them. I'm not generally in favor of taking lots of dependencies unless they're absolutely critical to the system working. I'd love to collect a list of recommended modules somewhere, maybe in the wiki?

> building a whole custom layer might be out of scope though

Um, yeah. I'm pretty sure we can do everything we need to without going to that amount of trouble. 😀

**Comment by rsek** [comment]
> > 
> 
> See #121 and #238 for the original impetus here, but there have been quite a few requests over the lifetime of this project. Especially when these items can represent named NPCs, it sometimes makes sense to plop them onto the map to remind you who your contact is on this planet, or where Admiral Carcosa's fleet is now. The use case isn't "gridded map for tactical combat," it's more "desktop that holds the minimized items that you're working with."

that's what i'm getting at, tho - can't *both* of those examples be accomplished with notes? 

like, as of v10:
* you can drag and drop journal entries to create a map note right out of the box
* it's automatically linked to a journal entry
* it's repositionable by dragging it with the mouse - currently only from the map note layer, but that's not set in stone any more than the baseline functionality of actors is set in stone
* the icon can be changed

that ticks all the boxes i can think of. is there some other function of actors that i'm missing that's important here?

there's some additional fringe benefits to *not* using actors for NPCs, too: 
* the  token hud won't be present - (IMO superfluous for ironsworn NPCs, unless a lot of work were put into modding it too; alternately, could just hide it?)

like, to put it another way: if we implement the journal entry page extension and make a tweak to map note draggability, then NPC map representation is pretty much done, because the maximum mechanical heft of an NPC is a progress track - they don't use any unique components. that's why i suggested seeing if people *need* full actors first: re-implementing and maintaining NPC actors might just be more work for no great benefit.

**Comment by ben** [comment]
> can't both of those examples be accomplished with notes?

Yes, but not well. Actors are a much better-developed thing in Foundry. And I've already outlined this clunky interaction mode. Pick pins mode, move your NPC, try to move your PC, remember that it's an actor, switch to actor mode, move it around, back to pins mode, move your NPC, try to double-click on your PC, remember it's an actor, back to actor mode… It's a paper cut every time you want to interact with it. The people who asked for "NPCs on the map" didn't have that in mind. Every time I try this out to see what it would be like to use it, I find another thing that I don't like about it. JE pins are hidden when you're in actor mode until you find the right sidebar button. JE pins don't show their titles unless you hover over them. JE pins can't emit funky animated light patterns.

> * the token hud won't be present

There's another one: JE pins can't have status badges. They can't rotate, either. Why should PCs get all these features and NPCs not?

> * that's not set in stone any more than the baseline functionality of actors is set in stone

The "not set in stone" phrase is doing a _lot_ of heavy lifting in that sentence. I mean we could make a spreadsheet with Foundry too. It's just code, right? 😉 

---

Look, it seems like we're really at loggerheads about the "just JE's" or "actors that reference JE's" thing. Let's put that on ice until I can whip up a demo and show you what I've got in mind. I suspect that "make it possible to drag map pins around while in actor mode" is both harder to get right and less desirable than it seems at first, and I really think that letting Foundry have its way with canvas layers is going to be better for us than trying to force one kind of entity to behave like another one.

**Comment by rsek** [comment]
> Look, it seems like we're really at loggerheads about the "just JE's" or "actors that reference JE's" thing. Let's put that on ice until I can whip up a demo and show you what I've got in mind. I suspect that "make it possible to drag map pins around while in actor mode" is both harder to get right and less desirable than it seems at first, and I really think that letting Foundry have its way with canvas layers is going to be better for us than trying to force one kind of entity to behave like another one.

yeah, that's fair. FWIW i think i'm basically with you at this point - i walked some of what i said back in later edits of my post before your reply, if you care to take a look :) sorry if you're replying based on an earlier version; i've left the original gist of it intact for historical reasons.

basically, i think that a modified actor token HUD might be leveraged to, say, provide a shortcut to an actor's associated journal entry, display progress, etc.

**Comment by ben** [comment]
A quick note to say that #508 is actually working now, so we can try it out and see how it feels.
